packages/eslint-plugin/src/rules/no-void-expression.ts

```diff
@@ -0,0 +1,286 @@
+      invalidVoidExprReturnLast:
+        'Returning a void expression from a function is forbidden. ' +
+        'Please remove the `return` statement.',
+      invalidVoidExprReturnWrapVoid:
+        'Void expressions returned from a function ' +
+        'must be marked explicitly with the `void` operator.',
+      voidExprWrapVoid: 'Mark with an explicit `void` operator',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreArrowShorthand: { type: 'boolean' },
+          ignoreVoidOperator: { type: 'boolean' },
+        },
+        additionalProperties: false,
+      },
+    ],
+    type: 'problem',
+    fixable: 'code',
+  },
+  defaultOptions: [{}],
+
+  create(context, [options]) {
+    return {
+      'AwaitExpression, CallExpression, TaggedTemplateExpression'(
+        node:
+          | TSESTree.AwaitExpression
+          | TSESTree.CallExpression
+          | TSESTree.TaggedTemplateExpression,
+      ): void {
+        const parserServices = util.getParserServices(context);
+        const checker = parserServices.program.getTypeChecker();
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+        const type = util.getConstrainedTypeAtLocation(checker, tsNode);
+        if (!tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike)) {
+          // not a void expression
+          return;
+        }
+
+        const invalidAncestor = findInvalidAncestor(node);
+        if (invalidAncestor == null) {
+          // void expression is in valid position
+          return;
+        }
+
+        const sourceCode = context.getSourceCode();
+        const wrapVoidFix = (fixer: TSESLint.RuleFixer): TSESLint.RuleFix => {
+          const nodeText = sourceCode.getText(node);
+          const newNodeText = `void ${nodeText}`;
+          return fixer.replaceText(node, newNodeText);
+        };
+
+        // handle arrow function shorthand
+        if (invalidAncestor.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+          // handle wrapping with `void`
+          if (options.ignoreVoidOperator) {
+            return void context.report({
+              node,
+              messageId: 'invalidVoidExprArrowWrapVoid',
+              fix: wrapVoidFix,
+            });
+          }
+
+          // handle wrapping with braces
+          const arrowFunction = invalidAncestor;
+          return void context.report({
+            node,
+            messageId: 'invalidVoidExprArrow',
+            fix(fixer) {
+              const arrowBody = arrowFunction.body;
+              const arrowBodyText = sourceCode.getText(arrowBody);
+              const newArrowBodyText = `{ ${arrowBodyText}; }`;
+              if (isParenthesized(arrowBody, sourceCode)) {
+                const [bodyOpeningParen, bodyClosingParen] = [
+                  sourceCode.getTokenBefore(arrowBody, isOpeningParenToken)!,
+                  sourceCode.getTokenAfter(arrowBody, isClosingParenToken)!,
+                ];
```

Comment:
nit - no need to assign an array here just to immediately destructure it
```suggestion
                const bodyOpeningParen = sourceCode.getTokenBefore(arrowBody, isOpeningParenToken)!;
                const bodyClosingParen = sourceCode.getTokenAfter(arrowBody, isClosingParenToken)!;
```

Reply:
I just thought it looks nicer